migrations/lib/importer/steps/tag_groups.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+module Migrations::Importer::Steps
+  class TagGroups < ::Migrations::Importer::CopyStep
+    MAX_NAME_LENGTH = 100
+
+    depends_on :tags
+    store_mapped_ids true
+
+    requires_mapping :existing_tag_group_by_name, "SELECT LOWER(name), id FROM tag_groups"
+
+    column_names %i[id name one_per_topic created_at updated_at parent_tag_id]
+
+    total_rows_query <<~SQL, MappingType::TAG_GROUPS
+      SELECT COUNT(*)
+      FROM tag_groups
+           LEFT JOIN mapped.ids mapped_tag_group
+              ON tag_groups.original_id = mapped_tag_group.original_id
+                 AND mapped_tag_group.type = ?
+      WHERE mapped_tag_group.original_id IS NULL
+    SQL
+
+    rows_query <<~SQL, MappingType::TAG_GROUPS, MappingType::TAGS
+      SELECT tag_groups.*,
+             mapped_tag.discourse_id AS discourse_parent_tag_id
+      FROM tag_groups
+           LEFT JOIN mapped.ids mapped_tag_group
+             ON tag_groups.original_id = mapped_tag_group.original_id
+                AND mapped_tag_group.type = ?1
+           LEFT JOIN mapped.ids mapped_tag
+             ON tag_groups.parent_tag_id = mapped_tag.original_id AND mapped_tag.type = ?2
+      WHERE mapped_tag_group.original_id IS NULL
+      ORDER BY tag_groups.original_id
+    SQL
+
+    def execute
+      SiteSetting.tags_listed_by_group = true
```

Comment:
Should we update that site setting only when there are tag groups? The `before` hook from https://github.com/discourse/discourse/pull/34491 could have the row count as a parameter.

Reply:
Yes, that works, even though there’s a scenario where all available rows could be skipped too. I think just checking whether there are any rows is good enough for now.